xalia/Win32/HwndTabProvider.cs

```diff
@@ -245,11 +274,72 @@ public override bool WatchProperty(UiDomElement element, GudlExpression expressi
                             Utils.RunTask(FetchSelectionIndex());
                         }
                         return true;
+                    case "win32_item_rects":
+                        watching_item_rects = true;
+                        if (!ItemRectsKnown)
+                        {
+                            Utils.RunTask(FetchItemRects());
+                        }
+                        return true;
                 }
             }
             return base.WatchProperty(element, expression);
         }
 
+        private async Task FetchItemRects()
+        {
+            int prev_change_count = item_rects_change_count;
+            RECT[] result = new RECT[ItemCount];
+            var mem = Win32RemoteProcessMemory.FromPid(Pid);
+            try {
+                using (var mem_result = mem.Alloc<RECT>())
+                {
+                    for (int i = 0; i < ItemCount; i++)
+                    {
+                        var msg_result = await SendMessageAsync(Hwnd, TCM_GETITEMRECT,
+                            new IntPtr(i), new IntPtr(unchecked((long)mem_result.Address)));
+                        if (item_rects_change_count != prev_change_count)
+                            return;
+                        if (msg_result == IntPtr.Zero)
+                            continue;
+                        result[i] = mem_result.Read<RECT>();
+                    }
+                }
+            }
+            finally
+            {
+                mem.Unref();
+            }
+            ItemRectsKnown = true;
+            ItemRects = result;
+            if (Element.MatchesDebugCondition())
+                Utils.DebugWriteLine($"{Element}.win32_item_rects: {new Win32ItemRects(result)}");
+            Element.PropertyChanged("win32_item_rects");
+        }
+
+        public void InvalidateItemRects()
+        {
+            item_rects_change_count++;
+            if (watching_item_rects)
+                Utils.RunTask(FetchItemRects());
+            else
+                ItemRectsKnown = false;

```

Comment:
Do we not want `ItemRectsKnown` here to be set to false in both cases? `FetchItemRects()` sets it to true, but this is the only place it gets set explicitly to false AFAICT. Or is it the case that when being watched, it's always expected to be known?

Reply:
Well, changing it to unknown means that attempts to access it are going to return `undefined`. If we do that while the value is not in use (not being watched), it doesn't matter. If the value is in use, then it transitions from <value 1> -> undefined -> <value 2>. I figured it was better to skip that intermediate step.

Reply:
Makes sense. Just wanted to make sure this was deliberate and not an oversight.

With that out of the way, this looks good to me.